### DISCUSSION TITLE
Custom Output Widget Issue

### DISCUSSION THREAD

**Discussion by maheshrajrp** [eclipse-theia/theia#10350] [Q&A] [answered]
Hi,
I'm trying to link OutputWidget to change colour in statusBar everytime it's opened/closed.

`    bind(CustomOutputWidget).toSelf().inSingletonScope();`
`    rebind(OutputWidget).toService(CustomOutputWidget); `

```
import { OutputWidget } from '@theia/output/lib/browser/output-widget';

injectable()
export class CustomOutputWidget extends OutputWidget {
    @inject(StatusBar) protected readonly statusBar: CustomStatusBar;


    protected onAfterShow(msg) {
        this.statusBar.onStatusBarElementActive(MESSAGE_WIDGET_ID)
        super.onAfterShow(msg);
    }

    protected onAfterHide(msg) {
        this.statusBar.onStatusBarElementClosed(MESSAGE_WIDGET_ID);
        super.onAfterHide(msg);
    }

    protected onCloseRequest(msg) {
        this.statusBar.onStatusBarElementClosed(MESSAGE_WIDGET_ID);
        super.onCloseRequest(msg);
    }

}
```

Everything works fine, except when the OutputWidget is closed. There some bug which turns the Tab upon reopening from this 
![image](https://user-images.githubusercontent.com/80001447/139362377-bc5f6c0d-685c-472b-8629-33ad59d0248e.png)

To
![image](https://user-images.githubusercontent.com/80001447/139361631-eb7dd741-63af-49c7-8082-bcc266e8251d.png)

As you can see the icon and text goes missing and the widget turns unsable. I'm unable to close it too. Using shortcut also results in vain. However, on reload it works fine until I close it again. We are also contributing `message-log` to it. Might it be the reason ?

Am I missing something ? Also, Am I importing the correct file? 
Any help would be appreciated.

Thanks

**Comment by vince-fugnitto** [comment, marked as answer]
@maheshrajrp thank you for the discussion, unfortunately it is difficult to help without seeing all of the code.
I'd suggest looking for any logs or errors (both in the [browser dev-console](https://developer.chrome.com/docs/devtools/open/#console) and backend (where the app is started)), and/or debug for yourself as to why it might not work properly.

**Reply by maheshrajrp** [reply]
Hi @vince-fugnitto ,
Thanks for your reply. There isnt any logs or errors regarding that. I was hoping if you could confirm the method I'm using rebind OutputWidget and overriding functionality is correct since that is the part I'm unsure. 

Thanks.

**Reply by vince-fugnitto** [reply]
@maheshrajrp I guess you can play around with it more, in my case I extended the default `OutputWidget` in the following way:

```ts
import { injectable, postConstruct } from '@theia/core/shared/inversify';
import { OutputWidget } from '@theia/output/lib/browser/output-widget';

@injectable()
export class CustomOutputWidget extends OutputWidget {

    constructor() {
        super();
        this.title.label = 'Custom Output';
    }

    @postConstruct()
    protected init(): void {
        super.init();
        console.warn('CustomOutputWidget was initialized.');
    }
}
```

and binded it like so:

```ts
import { ContainerModule } from '@theia/core/shared/inversify';
import { CustomOutputWidget } from './custom-output-widget';
import { OutputWidget } from '@theia/output/lib/browser/output-widget';
import { WidgetFactory } from '@theia/core/lib/browser';

export default new ContainerModule(bind => {
    bind(CustomOutputWidget).toSelf();
    bind(WidgetFactory).toDynamicValue(ctx => ({
        id: OutputWidget.ID, // use existing `OutputWidget` id.
        createWidget: () => ctx.container.get<CustomOutputWidget>(CustomOutputWidget)
    })).inSingletonScope();
});
```

The result is the following:

![image](https://user-images.githubusercontent.com/40359487/139443542-091d1da2-313f-4fbe-bf95-00e478105abf.png)

**Reply by maheshrajrp** [reply]
Hi @vince-fugnitto ,

Thanks for going wayy out. This is exactly what I needed. I just need to understand like how bindings works, and how to know which one to use. I'm currently working based on trial and error. It would really be great if you could point me at a direction to begin with so that I could fully understand. 

Thanks alot again.

**Reply by vince-fugnitto** [reply]
@maheshrajrp I simply looked at how the original binding was made where we use the `WidgetFactory` to return us an instance:

https://github.com/eclipse-theia/theia/blob/e5e6f928e39e88d51c4beb005ceb577fe0d3c26e/packages/output/src/browser/output-frontend-module.ts#L42-L45

I can then use the same way to bind the existing `id` to the new implementation.
For other binding related questions, you can take a look at [inversify](https://github.com/inversify/InversifyJS) itself and it's documentation.

**Reply by maheshrajrp** [reply]
Thanks a ton @vince-fugnitto . I'll check those.
